File: packages/vmcp-dsl/src/patterns/pipeline.ts
/**
 * Pipeline pattern builder
 */

import type {
  PatternSpec,
  PipelineSpec,
  PipelineStep,
  StepOperation,
  DataBinding,
  ConstructBinding,
} from '../types.js';
import { path } from '../path-builder.js';

/**
 * Builder for pipeline steps
 */
export class StepBuilder {
  private step: Partial<PipelineStep> = {};

  constructor(id: string) {
    this.step.id = id;
  }

  /**
   * Set step to call a tool
   */
  tool(name: string): this {
    this.step.operation = { tool: { name } };
    return this;
  }

  /**
   * Set step to execute a pattern
   */
  pattern(spec: PatternSpec): this {
    this.step.operation = { pattern: spec };
    return this;
  }

  /**
   * Set input from composition input
   */
  fromInput(pathExpr: string = '$'): this {
    this.step.input = { input: { path: path(pathExpr) } };
    return this;
  }

  /**
   * Set input from a previous step
   */
  fromStep(stepId: string, pathExpr: string = '$'): this {
    this.step.input = { step: { stepId, path: path(pathExpr) } };
    return this;
  }

  /**
   * Set input to a constant value
   */
  constant(value: unknown): this {
    this.step.input = { constant: value };
    return this;
  }

  /**
   * Construct input from multiple field bindings
   * Allows building a new object from fields extracted from prior steps or input
   */
  construct(fields: Record<string, DataBinding>): this {
    this.step.input = { construct: { fields } };
    return this;
  }

  /**
   * Build the step
   */
  build(): PipelineStep {
    if (!this.step.id) {
      throw new Error('Step ID is required');
    }
    if (!this.step.operation) {
      throw new Error('Step operation (tool or pattern) is required');
    }
    return this.step as PipelineStep;
  }
}

/**
 * Builder for pipeline patterns
 */
export class PipelineBuilder {
  private steps: PipelineStep[] = [];

  /**
   * Add a step that calls a tool
   */
  step(id: string, toolName: string): this {
    this.steps.push({
      id,
      operation: { tool: { name: toolName } },
    });
    return this;
  }

  /**
   * Add a step with full configuration
   */
  addStep(step: PipelineStep): this {
    this.steps.push(step);
    return this;
  }

  /**
   * Add a step using the step builder
   */
  add(id: string): StepBuilder & { then: () => PipelineBuilder } {
    const builder = new StepBuilder(id) as StepBuilder & { then: () => PipelineBuilder };
    builder.then = () => {
      this.steps.push(builder.build());
      return this;
    };
    return builder;
  }

  /**
   * Build the pipeline pattern spec
   */
  build(): PatternSpec {
    return { pipeline: { steps: this.steps } };
  }

  /**
   * Get the raw pipeline spec
   */
  spec(): PipelineSpec {
    return { steps: this.steps };
  }
}

/**
 * Create a pipeline pattern
 *
 * @example
 * const searchPipeline = pipeline()
 *   .step('search', 'web_search')
 *   .step('summarize', 'summarize_text')
 *   .build();
 *
 * // With step builder
 * const pipeline2 = pipeline()
 *   .add('search').tool('web_search').fromInput('$.query').then()
 *   .add('process').tool('process').fromStep('search', '$.results').then()
 *   .build();
 */
export function pipeline(): PipelineBuilder {
  return new PipelineBuilder();
}

/**
 * Create a step builder
 */
export function step(id: string): StepBuilder {
  return new StepBuilder(id);
}

// =============================================================================
// Binding Helper Functions
// =============================================================================

/**
 * Create an input binding (reference to composition input)
 */
export function fromInput(pathExpr: string = '$'): DataBinding {
  return { input: { path: path(pathExpr) } };
}

/**
 * Create a step binding (reference to prior step output)
 */
export function fromStep(stepId: string, pathExpr: string = '$'): DataBinding {
  return { step: { stepId, path: path(pathExpr) } };
}

/**
 * Create a constant binding
 */
export function constant(value: unknown): DataBinding {
  return { constant: value };
}

/**
 * Create a construct binding (build object from multiple field bindings)
 *
 * @example
 * construct({
 *   product_id: fromStep('alerts', '$.alerts[0].product_id'),
 *   quantity: fromStep('alerts', '$.alerts[0].deficit'),
 * })
 */
export function construct(fields: Record<string, DataBinding>): DataBinding {
  return { construct: { fields } };
}

File: crates/agentgateway/src/mcp/registry/patterns/pipeline.rs
// Pipeline pattern types

use serde::{Deserialize, Serialize};

use super::PatternSpec;

/// PipelineSpec executes steps sequentially, passing output to next step
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineSpec {
	/// Steps to execute in order
	pub steps: Vec<PipelineStep>,
}

impl PipelineSpec {
	/// Get the names of tools referenced by this pipeline
	pub fn referenced_tools(&self) -> Vec<&str> {
		self.steps
			.iter()
			.flat_map(|step| step.operation.referenced_tools())
			.collect()
	}
}

/// A single step in a pipeline
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PipelineStep {
	/// Unique identifier for this step (for data binding references)
	pub id: String,

	/// The operation to execute
	pub operation: StepOperation,

	/// Input binding for this step
	#[serde(default)]
	pub input: Option<DataBinding>,
}

/// StepOperation defines what a step does
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum StepOperation {
	/// Call a tool by name
	Tool(ToolCall),

	/// Inline pattern (no separate name)
	Pattern(Box<PatternSpec>),
}

impl StepOperation {
	/// Get the names of tools referenced by this operation
	pub fn referenced_tools(&self) -> Vec<&str> {
		match self {
			StepOperation::Tool(tc) => vec![tc.name.as_str()],
			StepOperation::Pattern(p) => p.referenced_tools(),
		}
	}
}

/// Tool call reference
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCall {
	/// Tool name (can be virtual tool, composition, or backend tool)
	pub name: String,
}

/// DataBinding specifies where step input comes from
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum DataBinding {
	/// From composition input
	Input(InputBinding),

	/// From a previous step's output
	Step(StepBinding),

	/// Constant value
	Constant(serde_json::Value),

	/// Construct an object from multiple bindings
	/// This enables input schema construction from prior step outputs
	Construct(ConstructBinding),
}

impl Default for DataBinding {
	fn default() -> Self {
		DataBinding::Input(InputBinding { path: "$".to_string() })
	}
}

/// Input binding - reference to composition input
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InputBinding {
	/// JSONPath into composition input
	pub path: String,
}

/// Step binding - reference to a previous step's output
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StepBinding {
	/// ID of the step to reference
	pub step_id: String,

	/// JSONPath into step output
	pub path: String,
}

/// Construct binding - build an object from multiple bindings
/// Enables symmetric input construction (like outputTransform does for outputs)
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConstructBinding {
	/// Field name -> binding that produces the field value
	pub fields: std::collections::HashMap<String, DataBinding>,
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_parse_simple_pipeline() {
		let json = r#"{
			"steps": [
				{
					"id": "search",
					"operation": { "tool": { "name": "web_search" } },
					"input": { "input": { "path": "$" } }
				},
				{
					"id": "summarize",
					"operation": { "tool": { "name": "summarize" } },
					"input": { "step": { "stepId": "search", "path": "$" } }
				}
			]
		}"#;

		let pipeline: PipelineSpec = serde_json::from_str(json).unwrap();
		assert_eq!(pipeline.steps.len(), 2);
		assert_eq!(pipeline.steps[0].id, "search");
		assert_eq!(pipeline.steps[1].id, "summarize");
	}

	#[test]
	fn test_parse_step_operation_tool() {
		let json = r#"{ "tool": { "name": "fetch" } }"#;
		let op: StepOperation = serde_json::from_str(json).unwrap();
		assert!(matches!(op, StepOperation::Tool(_)));
		if let StepOperation::Tool(tc) = op {
			assert_eq!(tc.name, "fetch");
		}
	}

	#[test]
	fn test_parse_data_binding_input() {
		let json = r#"{ "input": { "path": "$.query" } }"#;
		let binding: DataBinding = serde_json::from_str(json).unwrap();
		assert!(matches!(binding, DataBinding::Input(_)));
	}

	#[test]
	fn test_parse_data_binding_step() {
		let json = r#"{ "step": { "stepId": "step1", "path": "$.results" } }"#;
		let binding: DataBinding = serde_json::from_str(json).unwrap();
		assert!(matches!(binding, DataBinding::Step(_)));
	}

	#[test]
	fn test_parse_data_binding_constant() {
		let json = r#"{ "constant": "fixed_value" }"#;
		let binding: DataBinding = serde_json::from_str(json).unwrap();
		assert!(matches!(binding, DataBinding::Constant(_)));
	}

	#[test]
	fn test_referenced_tools() {
		let json = r#"{
			"steps": [
				{
					"id": "s1",
					"operation": { "tool": { "name": "tool_a" } }
				},
				{
					"id": "s2",
					"operation": { "tool": { "name": "tool_b" } }
				}
			]
		}"#;

		let pipeline: PipelineSpec = serde_json::from_str(json).unwrap();
		let refs = pipeline.referenced_tools();
		assert_eq!(refs, vec!["tool_a", "tool_b"]);
	}
}

